test/Facilitator/performProgressMint.js

```diff
@@ -0,0 +1,243 @@
+        '0x0000000000000000000000000000000000000000000000000000000000000001',
+      unlockSecret:
+        '0x0000000000000000000000000000000000000000000000000000000000000002',
+      txOptions: {
+        from: '0x0000000000000000000000000000000000000001',
+        gas: '7500000',
+      },
+    };
+    getInboxMessageStatusResult = MessageStatus.DECLARED;
+    progressMintResult = true;
+  });
+
+  it('should throw an error when message hash is undefined', async () => {
+    delete progressMintParams.messageHash;
+    await AssertAsync.reject(
+      facilitator.performProgressMint(
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ),
+      `Invalid message hash: ${progressMintParams.messageHash}.`,
+    );
+  });
+
+  it('should throw an error when unlock secret is undefined', async () => {
+    delete progressMintParams.unlockSecret;
+    await AssertAsync.reject(
+      facilitator.performProgressMint(
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ),
+      `Invalid unlock secret: ${progressMintParams.unlockSecret}.`,
+    );
+  });
+
+  it('should throw an error when transaction options is undefined', async () => {
+    delete progressMintParams.txOptions;
+    await AssertAsync.reject(
+      facilitator.performProgressMint(
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ),
+      `Invalid transaction option: ${progressMintParams.txOptions}.`,
+    );
+  });
+
+  it('should throw an error when from address in transaction options is undefined', async () => {
+    delete progressMintParams.txOptions.from;
+    await AssertAsync.reject(
+      facilitator.performProgressMint(
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ),
+      `Invalid from address ${
+        progressMintParams.txOptions.from
+      } in transaction options.`,
+    );
+  });
+
+  it('should throw an error when inbox message status is undeclared', async () => {
+    getInboxMessageStatusResult = MessageStatus.UNDECLARED;
+    setup();
+    await AssertAsync.reject(
+      facilitator.performProgressMint(
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ),
+      'Message cannot be progressed.',
+    );
+
+    SpyAssert.assert(spyGetInboxMessageStatus, 1, [
+      [progressMintParams.messageHash],
+    ]);
+    SpyAssert.assert(spyProgressMint, 0, [[]]);
+    SpyAssert.assert(spyCall, 1, [
+      [
+        progressMintParams.messageHash,
+        progressMintParams.unlockSecret,
+        progressMintParams.txOptions,
+      ],
+    ]);
+
+    teardown();
+  });
+
+  it('should throw an error when inbox message status is revocation declared', async () => {
```

Comment:
I think for facilitator it should know the status of the message so that it can resume the partially completed tasks.
If the state is not known and facilitator call progressStake, and it fails for confirmStakeIntent, how will facilitator know the reason? It can be known by the state only.
Also if the message is already progressed, it will fail if called again.
I prefer to provide a proper error message instead of failing it without giving a proper reason.

@schemar and @hobofan, do you also think to remove the logic and simply make contract interact calls? If yes then I can create a ticket for this.

Reply:
@deepesh-kn I agree. Even though I don't like duplicating it, I'm not sure if there is a way to provide good error messages without doing so...

Reply:
We have added error check logic on contract interact, it is also duplicating of the contract's logic. In that case, too we can just make a call and check to avoid transaction failure. 🤔 
Maybe we can discuss this in detail on call. I will be open to making changes.

Reply:
If we can get the error from the transaction (revert), then we can just forward that? Or am I missing something?